e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +214,95 @@ func (pf *PortForwarder) start() error {
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	err := pf.waitForFailure()
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		return fmt.Errorf("exiting - PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+	}
+	pf.logger.Infof("received signal for port-forward for %s/%s. err: %v", pf.ns, pf.pod, err)
+	if err := pf.portForwardToDeployment(true); err != nil {
+		// this is a potentially fatal error since we failed to restart the port
+		// forward. we log the error and return in case the test is already done
+		// with the port forward. the test will otherwise error the next time it
+		// tries to use the local port.
+		return err
+	}
+	return nil
+}
+
+// waitForFailure waits for any failure condition which indicates the port-forward
+// must be restarted. The port-forward must be restarted if:
+// - the subprocess is killed
+// - the pod is restarted
+func (pf *PortForwarder) waitForFailure() error {
+	ctx, cancel := context.WithCancel(pf.ctx)
+	cmdChan := functionWithChannel(func() error {
+		// cmd.Wait does not accept a context, so create a wrapper that cancels with the ctx
+		cmdWaitChan := functionWithChannel(pf.cmd.Wait)
+		var chanErr error
+		select {
+		case <-ctx.Done():
+			chanErr = ctx.Err()
+		case err := <-cmdWaitChan:
+			chanErr = err
+		}
+		// kill the subprocess. This should force the cmd.Wait goroutine to return, if it hasn't already.
+		if err := pf.cmd.Process.Kill(); err != nil && !errors.Is(err, os.ErrProcessDone) {
+			// potentially bad if this fails, cmd.Wait may never return?
+			pf.logger.Infof("failed to kill port forward process", err)
+		} else if err == nil {
+			pf.logger.Infof("stopped port-forward process for %s/%s", pf.ns, pf.pod)
+		}
+		// wait for cmdWaitChan to close, i.e. cmd.Wait returns
+		<-cmdWaitChan
+		return chanErr
+	})
+	podChan := functionWithChannel(func() error {
+		err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns,
+			testwatcher.WatchTimeout(0), testwatcher.WatchContext(ctx))
+		// We're using WatchForNotFound to detect if the Pod is deleted, but a nil
+		// error can be confusing in upstream logging. Adjust the returned error accordingly.
+		if err == nil {
+			// Pod not found, return an error
+			return fmt.Errorf("%s/%s Pod not found", pf.ns, pf.pod)
+		}
+		return err
+	})
+	defer func() {
+		// before we return, cancel and wait for goroutines to exit
+		cancel()
+		<-cmdChan
+		<-podChan
+	}()
+	// wait for any of the channels to return. a happy little race
+	select {
+	case err := <-cmdChan:
+		return err
+	case err := <-podChan:
+		return err
+	}
+}
+
+// functionWithChannel takes a blocking function and invokes it asynchronously in a
+// goroutine. A channel is returned which will be written to when the blocking
+// function returns.
+func functionWithChannel(fn func() error) chan error {
+	doneCh := make(chan error)
```

Comment:
nit: "doneCh" is usually used for `chan struct{}` that just close when done. This is more of an errorCh or resultCh, since it also sends a nil when there's no error. Speaking of which... should it? Or does that need an `if err != nil` check?

Reply:
renamed to `errorCh`